contracts/src/token/erc20/extensions/permit.rs

```diff
@@ -0,0 +1,196 @@
+        /// Nonces contract.
+        Nonces nonces;
+        /// EIP-712 contract. Must implement [`IEIP712`] trait.
+        T eip712;
+    }
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl<T: IEIP712 + StorageType> TopLevelStorage for Permit<T> {}
+
+#[external]
+impl<T: IEIP712 + StorageType> Permit<T> {
+    /// Returns the current nonce for `owner`.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `owner` - The address for which to return the nonce.
+    #[must_use]
+    pub fn nonces(&self, owner: Address) -> U256 {
+        self.nonces.nonces(owner)
+    }
+
+    /// Returns the domain separator used in the encoding of the signature for
+    /// [`Self::permit`], as defined by EIP712.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    #[selector(name = "DOMAIN_SEPARATOR")]
+    #[must_use]
+    pub fn domain_separator(&self) -> B256 {
+        self.eip712.domain_separator_v4()
+    }
+}
+
+impl<T: IEIP712 + StorageType> Permit<T> {
+    /// Sets `value` as the allowance of `spender` over `owner`'s tokens,
+    /// given `owner`'s signed approval.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state. given address.
+    /// * `owner` - Account that owns the tokens.
+    /// * `spender` - Account that will spend the tokens.
+    /// * `value` - The number of tokens being permitted to transfer by
+    ///   `spender`.
+    /// * `deadline` - Deadline for the permit action.
+    /// * `v` - v value from the `owner`'s signature.
+    /// * `r` - r value from the `owner`'s signature.
+    /// * `s` - s value from the `owner`'s signature.
+    /// * `erc20` - Write access to a contract providing
+    ///   [`crate::token::erc20::IErc20`] interface.
+    ///
+    /// # Errors
+    ///
+    /// If the `deadline` param is from the past, than the error
+    /// [`ERC2612ExpiredSignature`] is returned.
+    /// If signer is not an `owner`, than the error
+    /// [`ERC2612InvalidSigner`] is returned.
+    ///
+    /// # Events
+    ///
+    /// Emits an [`crate::token::erc20::Approval`] event.
+    ///
+    /// # Requirements
+    ///
+    /// * `spender` cannot be the ``Address::ZERO``.
+    /// * `deadline` must be a timestamp in the future.
+    /// * `v`, `r` and `s` must be a valid secp256k1 signature from `owner`
+    /// over the EIP712-formatted function arguments.
+    /// * the signature must use `owner`'s current nonce.
+    #[allow(clippy::too_many_arguments)]
+    pub fn permit(
+        &mut self,
+        owner: Address,
+        spender: Address,
+        value: U256,
+        deadline: U256,
+        v: u8,
+        r: B256,
+        s: B256,
+        erc20: &mut impl IErc20Internal,
```

Comment:
Do you think we could implement this differently? I don't think the API should be like this.

Reply:
Basically it looks similar to [Erc721Enumerable::_add_token_to_owner_enumeration](https://github.com/OpenZeppelin/rust-contracts-stylus/blob/22ea8d6606137809f880deb50f2a0cd76f720c05/contracts/src/token/erc721/extensions/enumerable.rs#L157). Since we accepted this approach once.. 
But yeah. Tough to say what should be better for users. Since stylus-sdk is quite limited for smart contract's extensions.

Reply:
Resolved with #235